src/awkward/_backends/backend.py

```diff
@@ -34,14 +35,35 @@ def index_nplike(self) -> NumpyLike:
     def __getitem__(self, key: KernelKeyType) -> KernelType:
         raise NotImplementedError
 
-    def apply_reducer(
-        self,
-        reducer: ak._reducers.Reducer,
-        layout: ak.contents.NumpyArray,
-        parents: ak.index.Index,
-        outlength: int,
-    ) -> ak.contents.NumpyArray:
-        return reducer.apply(layout, parents, outlength)
+    def prepare_reducer(self, reducer: ak._reducers.Reducer) -> ak._reducers.Reducer:
+        return reducer
 
     def apply_ufunc(self, ufunc, method, args, kwargs):
         return getattr(ufunc, method)(*args, **kwargs)
+
+    def format_kernel_error(
+        self,
+        error: KernelError,
+    ):
+        if error.filename is None:
+            filename = ""
+        else:
+            filename = " (in compiled code: " + error.filename.decode(
+                errors="surrogateescape"
+            ).lstrip("\n").lstrip("(")
+
+        message = error.str.decode(errors="surrogateescape")
+
+        assert not error.pass_through
```

Comment:
See #2504 for why this assertion holds

Reply:
Yes, it's dead code. The kernels part of Awkward had to be valid for both v1 and v2 at the same time. The error-reporting record is now fairly out of date, but we'd have to change it carefully, if at all. I don't remember whether the Numba implementation depends on that exact struct or not.